applications/models.py

```diff
@@ -1,3 +1,132 @@
+from django.contrib.auth.models import User
+from django.utils.translation import gettext_lazy as _
+from rest_framework.exceptions import ValidationError
+
+def not_before_now(year: int):
+    current_date = datetime.datetime.now()
+
+    if current_date.year < year:
+        raise ValidationError(
+            _("%(year)s is before current year"), params={"year": year}
+        )
+
+
+class Address(models.Model):
+    street_address = models.TextField(
+        verbose_name=_("Street address"), null=False, blank=False, max_length=80
+    )
+
+    post_code = models.PositiveIntegerField(
+        verbose_name=_("Post code"),
+        null=False,
+        blank=False,
+    )
+
+    city = models.TextField(
+        verbose_name=_("City"), null=False, blank=False, max_length=80
+    )
+
+
+class ContactInformation(models.Model):
+    email = models.EmailField(
+        verbose_name=_("Email"),
+        null=True,
+        blank=True,
+    )
+
+    phone_number = models.TextField(
+        verbose_name=_("Phone number"),
+        null=True,
+        blank=True,
+        max_length=50,
+    )
+
+    class Meta:
+        abstract = True
+
+
+class Person(ContactInformation):
+
+    first_name = models.TextField(
+        verbose_name=_("First name"), null=False, blank=False, max_length=50
+    )
+
+    last_name = models.TextField(
+        verbose_name=_("Last name"), null=False, blank=False, max_length=50
+    )
+
+
+class Organisation(models.Model):
+
+    name = models.TextField(
+        verbose_name=_("Name"),
+        null=False,
+        blank=False,
+        max_length=255,
+    )
+
+    identifier = models.TextField(
+        verbose_name=_("Identifier"),
+        null=False,
+        blank=False,
+        max_length=255,
+        unique=True,
+    )
+
+    year_established = (
+        models.PositiveIntegerField(
+            verbose_name=_("Year established"), validators=[not_before_now]
+        ),
+    )
+
+    address = models.ForeignKey(
+        Address, null=True, blank=True, on_delete=models.SET_NULL
+    )
+
+
+class Applicant(models.Model):
+
+    organisation = models.ForeignKey(
+        Organisation,
+        verbose_name=_("Organisation"),
+        null=False,
+        blank=False,
+        on_delete=models.PROTECT,
+    )
+
+    contact_person = models.ForeignKey(
+        Person,
+        verbose_name=_("Contact person"),
+        null=False,
+        blank=False,
+        on_delete=models.PROTECT,
+    )
+
+    user = models.ForeignKey(
+        User,
+        verbose_name=_("Applicant"),
+        null=False,
+        blank=False,
+        on_delete=models.PROTECT,
+    )
+
+    class Meta:
+        abstract = True
+
+
+class Application(Applicant):
+
+    description = models.fields.TextField(
+        verbose_name=_("Description"), max_length=1000, blank=True, null=True
+    )
 
-# Create your models here.
+    reservation_purpose = models.ForeignKey(
+        to="reservations.ReservationPurpose",
```

Comment:
This feels a bit hacky, with regular import there is a circular dep. Is there better way to handle this, like should these common stuff be in separate module or is this ok?